### PR TITLE
Exercise 2 group 2

### PR DESCRIPTION
**Pull request by zopad**
Could've used a bit more time to clean up the many ifs :)

### CODE REVIEW

File: Java/src/main/java/com/gildedrose/StoredItem.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        return this.name.equals(name);
+    }
+
+    @Override
+    public boolean isNotNamed(String name) {
+        return !this.name.equals(name);
+    }
+
+    @Override
+    public Item convertBack() {
```
Comment (corebonts) [inline Java/src/main/java/com/gildedrose/StoredItem.java:39]: StoredItem should be compatible with Item and needs no conversion.

File: Java/src/main/java/com/gildedrose/GildedRose.java [hunk 2]
```diff
@@ -9,54 +9,57 @@ public GildedRose(Item[] items) {
...
-                        if (items[i].quality > 0) {
-                            if (!items[i].name.equals("Sulfuras, Hand of Ragnaros")) {
-                                items[i].quality = items[i].quality - 1;
-                            }
-                        }
-                    } else {
-                        items[i].quality = items[i].quality - items[i].quality;
-                    }
+            if (currentItem.sellIn < 0) {
+                if (currentItem.isNamed("Aged Brie") && isLowQuality(currentItem)) {
```
Comment (corebonts) [inline Java/src/main/java/com/gildedrose/GildedRose.java:40]: why isLowQuality() is not part of the item then?

File: Java/src/main/java/com/gildedrose/GildedRose.java [hunk 3]
```diff
@@ -9,54 +9,57 @@ public GildedRose(Item[] items) {
...
+                                currentItem.increaseQuality();
                             }
                         }
                     }
                 }
             }
 
-            if (!items[i].name.equals("Sulfuras, Hand of Ragnaros")) {
-                items[i].sellIn = items[i].sellIn - 1;
+            if (currentItem.isNotNamed("Sulfuras, Hand of Ragnaros")) {
```
Comment (TehMillhouse) [inline Java/src/main/java/com/gildedrose/GildedRose.java:35]: This could have been pulled to the top

Comment (corebonts) [inline Java/src/main/java/com/gildedrose/GildedRose.java:35]: For sulfuras there's nothing to do, maybe an early return / continue would be more understandable.

File: Java/src/main/java/com/gildedrose/StoredItem.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        return new Item(this.name, this.sellIn, this.quality);
+    }
+    
+}
+
+interface NamedItem {
```
Comment (TehMillhouse) [inline Java/src/main/java/com/gildedrose/StoredItem.java:45]: Do these interfaces really make the brunt of the code easier to understand?

File: Java/src/main/java/com/gildedrose/GildedRose.java [hunk 5]
```diff
@@ -9,54 +9,57 @@ public GildedRose(Item[] items) {
...
     }
+
+    private boolean itemCanLoseQuality(StoredItem item) {
+        return item.quality > 0 && item.isNotNamed("Aged Brie")
+                && item.isNotNamed("Backstage passes to a TAFKAL80ETC concert")
+                && item.isNotNamed("Sulfuras, Hand of Ragnaros");
+    }
+
+    private boolean isLowQuality(StoredItem item) {
+        return item.quality < 50;
```
Comment (TehMillhouse) [inline Java/src/main/java/com/gildedrose/GildedRose.java:63]: Wow, that's quite a high standard for "low quality" :D

File: Java/src/main/java/com/gildedrose/GildedRose.java [hunk 6]
```diff
@@ -9,54 +9,57 @@ public GildedRose(Item[] items) {
...
-                    if (!items[i].name.equals("Backstage passes to a TAFKAL80ETC concert")) {
-                        if (items[i].quality > 0) {
-                            if (!items[i].name.equals("Sulfuras, Hand of Ragnaros")) {
-                                items[i].quality = items[i].quality - 1;
-                            }
-                        }
-                    } else {
-                        items[i].quality = items[i].quality - items[i].quality;
-                    }
+            if (currentItem.sellIn < 0) {
```
Comment (TehMillhouse) [inline Java/src/main/java/com/gildedrose/GildedRose.java:39]: I feel like the control flow could be cleaned up more here

File: Java/src/main/java/com/gildedrose/GildedRose.java [hunk 7]
```diff
@@ -9,54 +9,57 @@ public GildedRose(Item[] items) {
...
-                    items[i].quality = items[i].quality + 1;
+                if (isLowQuality(currentItem)) {
+                    currentItem.increaseQuality();
 
-                    if (items[i].name.equals("Backstage passes to a TAFKAL80ETC concert")) {
-                        if (items[i].sellIn < 11) {
-                            if (items[i].quality < 50) {
-                                items[i].quality = items[i].quality + 1;
+                    if (currentItem.isNamed("Backstage passes to a TAFKAL80ETC concert")) {
+                        if (currentItem.sellIn < 11) {
```
Comment (TehMillhouse) [inline Java/src/main/java/com/gildedrose/GildedRose.java:20]: these if blocks could be collapsed into one

File: Java/src/main/java/com/gildedrose/GildedRose.java [hunk 8]
```diff
@@ -9,54 +9,57 @@ public GildedRose(Item[] items) {
...
-                if (items[i].quality < 50) {
-                    items[i].quality = items[i].quality + 1;
+                if (isLowQuality(currentItem)) {
+                    currentItem.increaseQuality();
 
-                    if (items[i].name.equals("Backstage passes to a TAFKAL80ETC concert")) {
-                        if (items[i].sellIn < 11) {
-                            if (items[i].quality < 50) {
-                                items[i].quality = items[i].quality + 1;
+                    if (currentItem.isNamed("Backstage passes to a TAFKAL80ETC concert")) {
```
Comment (The-Drummer6) [inline Java/src/main/java/com/gildedrose/GildedRose.java:19]: that many if clauses make it hard to understand the code

File: Java/src/main/java/com/gildedrose/StoredItem.java (hunk 1 above)
Comment (zopad) [reply Java/src/main/java/com/gildedrose/StoredItem.java:39]: yep, brainfart

**Pull request merged**